# Wrangling Genomics

Lesson for quality control and wrangling genomics data. This repository is maintained by [Josh Herr](https://github.com/jrherr), [Ming Tang](https://github.com/crazyhottommy), [Lex Nederbragt](https://github.com/lexnederbragt), and [Fotis Psomopoulos](https://github.com/fpsom).

## Background

Wrangling genomics trains novice learners on a variant calling workflow. Participants will learn how to evaluate sequence quality and what to do if it is not good. We will then cover aligning reads to a genome, and calling variants, as well as discussing different file formats. Results will be visualized. Finally, we will cover how to automate the process by building a shell script.

This lesson is part of the [Data Carpentry](http://www.datacarpentry.org/) [Genomics Workshop](http://www.datacarpentry.org/genomics-workshop/).

## Self-study

If you want to go through this lesson on your own, there are several options:
- using a [cloud platform](http://www.datacarpentry.org/cloud-genomics/discuss/), there is an Amazon Machine Image (AMI) available with the data and software pre-installed on Amazon Web Services, [see the instructions](http://www.datacarpentry.org/cloud-genomics/discuss/#launching-an-instance-on-amazon-web-services)
- The trimmed fastq files, the corresponding reference genome and the shell script for variant analysis are available [from here](https://github.com/datacarpentry/wrangling-genomics/blob/gh-pages/variant_calling.tar.gz)

## Contribution

- Make a suggestion or correct an error by [raising an Issue](https://github.com/datacarpentry/wrangling-genomics/issues).

## Code of Conduct

All participants should agree to abide by the [Data Carpentry Code of Conduct](http://www.datacarpentry.org/code-of-conduct/).

## Authors

Wrangling genomics is authored and maintained by the [community](https://github.com/datacarpentry/wrangling-genomics/network/members).

## Citation

Please cite as:

Wrangling Genomics. June 2017. http://datacarpentry.github.io/wrangling-genomics/.
