# Wrangling Genomics

Lesson for quality control and wrangling genomics data. This repository is maintained by [Luis Avila](https://github.com/lmavila), and [Niel Infante](https://github.com/NielInfante). 

Amazon public AMI for this tutorial is "dataCgen-qc". 

## Background

Wrangling genomics trains novice learners on a variant calling workflow. Participants will learn how to evaluate sequence quality and what to do if it is not good. We will then cover aligning reads to a genome, and calling variants, as well as discussing different file formats. Results will be visualized. Finally, we will cover how to automate the process by building a shell script.

It builds on the work of [Software Carpentry](http://software-carpentry.org/) and [Data Carpentry](http://www.datacarpentry.org/).


## Contribution

There are many ways of contributing to Library Carpentry:

- Join our [![Gitter chat](https://badges.gitter.im/data-lessons/wrangling-genomics.png)](https://gitter.im/wrangling-genomics/Lobby)
- Follow updates on our *soon to be active* [Twitter](https://twitter.com/WranglingGenomics).
- Make a suggestion or correct an error by [raising an Issue](https://github.com/data-lessons/wrangling-genomics/issues).

## Code of Conduct

All participants should agree to abide by the [Software Carpentry Code of Conduct](http://software-carpentry.org/conduct/).

## Authors

Wrangling genomics is authored and maintained by the [community](https://github.com/data-lessons/wrangling-genomics/network/members).

## Citation

Please cite as:

Wrangling Genomics. June 2017. http://data-lessons.github.io/wrangling-genomics/.
